### ISSUE TITLE
miniFlow lesson 6 - SGD solution

### ISSUE THREAD

**Issue by bdholt1** [udacity/sdc-issue-reports#142]
The SGD solution is overly complicated:

        # Performs SGD
        # Get a list of the partials with respect to each trainable input.
        partials = [n.gradients[n] for n in trainables]
        # Loop over the trainables
        for n in range(len(trainables)):
            # Change the trainable's value by subtracting the learning rate
            # multiplied by the partial of the cost with respect to this
            # trainable.
            trainables[n].value -= learning_rate * partials[n]

This was my solution and I think it maps much better to the maths:

    for t in trainables:
                t.value -= learning_rate * t.gradients[t]

Also, I think this is just GD, and not SGD.  
>Stochastic Gradient Descent (SGD) is a version of Gradient Descent where a subset of the total dataset is randomly sampled and used (also known as a batch). 

Very clearly, the whole dataset is used in the batch, not just a small random sample.

**Comment by domluna** [comment]
See #113 

> Very clearly, the whole dataset is used in the batch, not just a small random sample.

Are you referring to the quiz? In the quiz entire dataset is used because it's so small. Hmm taking second look at the wording it could be interpreted as we select and subset of the data and when randomly sample from that, which is incorrect. I'll update the wording to make this more clear.

**Comment by domluna** [comment]
"Stochastic Gradient Descent (SGD) is a version of Gradient Descent where on each forward pass a batch of data is randomly sampled from total dataset."

Does this make more sense?

**Comment by domluna** [comment]
Updated the section, If you think there's still an issue reopen.

**Comment by bdholt1** [comment]
@domluna I was just saying that the section is called SGD, there is an explanation about what SGD is, but the code doesn't actually *do* SGD.  Obviously that's because the sample size is so small, but I fear the student won't understand what SGD is (because there's no code doing random subsampling), or will be confused (wondering why there's code doing random subsampling).  

Perhaps a solution might be to put the code in to do it but to note in a comment that the training set is too small to turn it on.

**Comment by domluna** [comment]
I understand. I bring it up a couple of times in the text that in the quiz we don't do random subsampling because the dataset is small enough it's not required.
